classes/setup_item_processor.php

```diff
@@ -81,78 +85,192 @@ public function process(setup_item $item) {
         }
 
         foreach ($courses as $course) {
-            $page = new \moodle_page();
-            $page->set_course($course);
-            $page->set_pagelayout('course');
-            $page->set_pagetype('course-view-' . $course->format);
-
-            if (!$page->blocks instanceof block_manager) {
-                // TODO: disable whole plugin if block manager is not overridden.
-                throw new \coding_exception('Terminate processing. Block manager class is not configured in config.php');
-            }
 
-            try {
-                $page->blocks->add_region($item->get_region(), false);
-                $page->blocks->load_blocks(true);
-                $page->blocks->create_all_block_instances();
+            $pagetypesegments = explode('-', $item->get_pagetypepattern());
+
+            // Check if block is to be added to a course module.
+            if ($pagetypesegments[0] == 'mod') {
```

Comment:
It feels like there there is a code duplication for mod processing and course processing. have you thought abut refactoring to avoid it? I'd suggest to start splitting this process method  to methods like process_mod  and process_course and then go down to those methods and see if you can abstract duplicated code to more methods and reuse.

Reply:
This has been resolved. There were three main parts that could be extracted:
1. checking if a block exists
2. adding a block
3. updating a block

We now have three new functions that encapsulate these functions that can be used in the new process_module and process_course methods, and any other possible extensions to the plugin.